### PR TITLE
Implementation of DBSCAN model

### PR DESCRIPTION
**Pull request by Kelang-Tian**
Implementation of the clustering model.
The test process will be submitted next time.

### CODE REVIEW

File: tests/test_dbscan.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+import logging
+from pathlib import Path
+from numpy import ndarray, testing
+
+iris = datasets.load_iris()
+iris_data = np.array(iris.data)
+iris_target = iris.target
+
+
+def check_model_exist(path):
```
Comment (Yancey0623) [inline tests/test_dbscan.py]: It seems that this function is not used?

File: Makefile [hunk 2]
```diff
@@ -6,7 +6,7 @@ setup: ## Setup virtual environment for local development
 	&& $(MAKE) install-requirements
 
 install-requirements:
-	pip install -U -e .
+	pip install --default-timeout=1000 -U -e .
```
Comment (Yancey0623) [inline Makefile:9]: Why increasing the timeout threshold ?

File: sqlflow_models/dbscan.py [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+__create_time__ : 2020/07/01
+"""
+import numpy as np
+import tensorflow as tf
+from scipy.spatial.distance import pdist, squareform
+from sklearn.base import BaseEstimator, ClusterMixin
+import pandas as pd
+from sklearn import datasets, metrics
+
+def optimizer():
```
Comment (Yancey0623) [inline sqlflow_models/dbscan.py:30]: It seems that it's not a NN model, should we remove this function?

File: sqlflow_models/dbscan.py [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+        self.clusters = self.generate_clusters(neighbors)
+        # filter out clusters with neighbors < minpts
+        self.clusters = list(filter(lambda x: len(x) >= self.minpts,
+                                    self.clusters))
+        # labelling
+        self.labels_ = np.array(self.labelling(X, self.clusters))
+
+        return self
+
+    def _split_dataset(self, dataset):
```
Comment (Yancey0623) [inline sqlflow_models/dbscan.py]: Do we actually need this function?

File: sqlflow_models/dbscan.py [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+        return clusters
+
+    def labelling(self, data, clusters):
+        cluster_labels = [self.OUTLIER] * len(data)
+        for i in range(len(self.clusters)):
+            for j in range(len(self.clusters[i])):
+                cluster_labels[self.clusters[i][j]] = i
+        return cluster_labels
+
+    def fit(self, X):
```
Comment (Yancey0623) [inline sqlflow_models/dbscan.py]: Instead of overriding `fit` function, we can implement `sqlflow_train_loop` to define the custom train loop logic. You may find that `sqlflow_train_loop` is not the standard API of Keras model, it's just let SQLFlow `runtime` to know to run the custom train loop.

c.f. https://github.com/sql-machine-learning/sqlflow/blob/9628487783878e9f019b1bf379e192134344e3a3/python/sqlflow_submitter/tensorflow/train_keras.py#L173

File: sqlflow_models/dbscan.py [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+            print("DBSCAN (minpts=10, eps=0.4): %f" %
+                  purity_score(y_df, self.labels_))
+        else:
+            x_df, y_df = self._split_dataset(dataset)
+            self.fit_predict(x_df)
+            print("DBSCAN (minpts=10, eps=0.4): %f" %
+                  purity_score(y_df, self.labels_))
+'''
+if __name__ == '__main__':
+    from sklearn.datasets.samples_generator import make_blobs
```
Comment (Yancey0623) [inline sqlflow_models/dbscan.py]: Can we remove the comment code and testing this model in the `test_db_scan.py`? A referenced test script: https://github.com/sql-machine-learning/models/blob/develop/tests/test_arima_with_stl_decomposition.py

File: tests/test_dbscan.py (hunk 1 above)
Comment (Kelang-Tian) [reply tests/test_dbscan.py]: Remove the function.

File: Makefile (hunk 2 above)
Comment (Kelang-Tian) [reply Makefile:9]: Timeout occurred while installing the packages because of network instability.

File: sqlflow_models/dbscan.py (hunk 3 above)
Comment (Kelang-Tian) [reply sqlflow_models/dbscan.py:30]: Yes, there is no need to set optimizer.

File: sqlflow_models/dbscan.py (hunk 4 above)
Comment (Kelang-Tian) [reply sqlflow_models/dbscan.py]: Replace this function with _read_Dataset_data, which split tf.dataset type data into features and labels(If it exists).

File: sqlflow_models/dbscan.py (hunk 5 above)
Comment (Kelang-Tian) [reply sqlflow_models/dbscan.py]: Enables `sqlflow_train_loop` function to handle tf.dataset type of data, because `sqlflow` server  invokes tf.dataset.
[https://github.com/sql-machine-learning/sqlflow/blob/483b8676cf93f373d5073d84b0bee311bb122012/python/runtime/tensorflow/input_fn.py#L72](url)

File: sqlflow_models/dbscan.py (hunk 6 above)
Comment (Kelang-Tian) [reply sqlflow_models/dbscan.py]: Comment code removed and rename the test_db_scan.py.
